core/commonMain/src/kotlinx/serialization/encoding/Encoding.kt

```diff
@@ -486,3 +486,24 @@ public inline fun Encoder.encodeStructure(descriptor: SerialDescriptor, block: C
         if (ex == null) composite.endStructure(descriptor)
     }
 }
+
+/**
+ * Begins a collection, encodes it using the given [block] and ends it.
+ */
+public inline fun Encoder.encodeCollection(
+    descriptor: SerialDescriptor,
+    collectionSize: Int,
+    block: CompositeEncoder.() -> Unit
+) {
+    val composite = beginCollection(descriptor, collectionSize)
+    var ex: Throwable? = null
+    try {
+        composite.block()
+    } catch (e: Throwable) {
+        ex = e
+        throw e
+    } finally {
+        // End structure only if there is no exception, otherwise it can be swallowed
+        if (ex == null) composite.endStructure(descriptor)
+    }
+}
```

Comment:
Given the way that collections work, it would make sense to put the element iteration in the helper, not just the structure part. Why not use something like:
```kotlin
fun encodeCollection(
    descriptor: SerialDescriptor,
    collection: Collection<T>,
    encodeElement: CompositeEncoder.(T) -> Unit
)
```

Reply:
You can encode things using a collection structure that aren't collections themselves. The simplest example are arrays.

I don't understand the first part, are you talking about which file the extension lives in or something else?

Reply:
@ansman The first part refers to the idea that you would possibly want `encodeCollection` to actually enforce/provide the expected semantics of a collection rather than leaving this to the caller. As to arrays, you can always overload the helper for array types.

Reply:
I like that idea but I don't want to reduce the flexibility of not enforcing a collection. Instead two extensions could be added (one that accepts a collection size and one that accepts a collection).